From 3046dd722c339f9b8b58c098033bb1551ff3747c Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:04:29 -0400
Subject: [fs] btrfs: tree_search: eliminate redundant nr_items check

Message-id: <1407179109-13714-383-git-send-email-sandeen@redhat.com>
Patchwork-id: 87324
O-Subject: [RHEL7.1 PATCH 382/422] btrfs: tree_search: eliminate redundant nr_items check
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Gerhard Heift <gerhard@heift.name>

commit 25c9bc2e2bc9d67de7fc49ff2784cae5de755192
Author: Gerhard Heift <gerhard@heift.name>
Date:   Thu Jan 30 16:23:57 2014 +0100

    btrfs: tree_search: eliminate redundant nr_items check

    If the amount of items reached the given limit of nr_items, we can leave
    copy_to_sk without updating the key. Also by returning 1 we leave the loop in
    search_ioctl without rechecking if we reached the given limit.

    Signed-off-by: Gerhard Heift <Gerhard@Heift.Name>
    Signed-off-by: Chris Mason <clm@fb.com>
    Acked-by: David Sterba <dsterba@suse.cz>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/ioctl.c b/fs/btrfs/ioctl.c
index c2e796b..65148cd 100644
--- a/fs/btrfs/ioctl.c
+++ b/fs/btrfs/ioctl.c
@@ -1995,7 +1995,7 @@ static noinline int copy_to_sk(struct btrfs_root *root,
   if (sizeof(sh) + item_len + *sk_offset >
       BTRFS_SEARCH_ARGS_BUFSIZE) {
    ret = 1;
-   goto overflow;
+   goto out;
   }
 
   sh.objectid = key->objectid;
@@ -2017,8 +2017,10 @@ static noinline int copy_to_sk(struct btrfs_root *root,
   }
   (*num_found)++;
 
-  if (*num_found >= sk->nr_items)
-   break;
+  if (*num_found >= sk->nr_items) {
+   ret = 1;
+   goto out;
+  }
  }
 advance_key:
  ret = 0;
@@ -2033,7 +2035,7 @@ advance_key:
   key->objectid++;
  } else
   ret = 1;
-overflow:
+out:
  return ret;
 }
 
@@ -2085,7 +2087,7 @@ static noinline int search_ioctl(struct inode *inode,
   ret = copy_to_sk(root, path, &key, sk, args->buf,
      &sk_offset, &num_found);
   btrfs_release_path(path);
-  if (ret || num_found >= sk->nr_items)
+  if (ret)
    break;
 
  }
-- 
1.7.1